File: packages/b2c-tooling-sdk/src/config/sources/mobify-source.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Mobify (~/.mobify) configuration source for MRT API key.
 *
 * @internal This module is internal to the SDK. Use ConfigResolver instead.
 */
import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import type {ConfigSource, ConfigLoadResult, ResolveConfigOptions} from '../types.js';
import {getLogger} from '../../logging/logger.js';

/**
 * Mobify config file structure (~/.mobify)
 */
interface MobifyConfigFile {
  username?: string;
  api_key?: string;
}

/**
 * Configuration source that loads MRT API key from ~/.mobify file.
 *
 * The mobify config file is a JSON file located at ~/.mobify containing:
 * ```json
 * {
 *   "username": "user@example.com",
 *   "api_key": "your-api-key"
 * }
 * ```
 *
 * When a cloudOrigin is provided in options, looks for ~/.mobify--[hostname] instead.
 *
 * @internal
 */
export class MobifySource implements ConfigSource {
  readonly name = 'MobifySource';
  readonly priority = 0;

  load(options: ResolveConfigOptions): ConfigLoadResult | undefined {
    const logger = getLogger();

    // Use explicit credentialsFile if provided, otherwise use default path
    const mobifyPath = options.credentialsFile ?? this.getMobifyPath(options.cloudOrigin);

    logger.trace({location: mobifyPath}, '[MobifySource] Checking for credentials file');

    if (!fs.existsSync(mobifyPath)) {
      logger.trace({location: mobifyPath}, '[MobifySource] No credentials file found');
      return undefined;
    }

    try {
      const content = fs.readFileSync(mobifyPath, 'utf8');
      const config = JSON.parse(content) as MobifyConfigFile;

      if (!config.api_key) {
        logger.trace({location: mobifyPath}, '[MobifySource] Credentials file found but no api_key present');
        return undefined;
      }

      logger.trace({location: mobifyPath, fields: ['mrtApiKey']}, '[MobifySource] Loaded credentials');

      return {
        config: {mrtApiKey: config.api_key},
        location: mobifyPath,
      };
    } catch (error) {
      // Invalid JSON or read error
      const message = error instanceof Error ? error.message : String(error);
      logger.trace({location: mobifyPath, error: message}, '[MobifySource] Failed to parse credentials file');
      return undefined;
    }
  }

  /**
   * Determines the mobify config file path based on cloud origin.
   */
  private getMobifyPath(cloudOrigin?: string): string {
    if (cloudOrigin) {
      // Extract hostname from origin URL for the config file suffix
      try {
        const url = new URL(cloudOrigin);
        return path.join(os.homedir(), `.mobify--${url.hostname}`);
      } catch {
        // If URL parsing fails, use the origin as-is
        return path.join(os.homedir(), `.mobify--${cloudOrigin}`);
      }
    }
    return path.join(os.homedir(), '.mobify');
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/config/sources/mobify_source.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Mobify (``~/.mobify``) configuration source for the MRT API key.

Mirrors ``src/config/sources/mobify-source.ts``. Reads the JSON ``~/.mobify``
credentials file and extracts ``api_key`` into ``mrt_api_key``. When a cloud
origin is provided, reads ``~/.mobify--<hostname>`` instead.
"""

from __future__ import annotations

import json as json_module
import os
from pathlib import Path
from urllib.parse import urlparse

from b2c_tooling_sdk.config.types import ConfigLoadResult, NormalizedConfig, ResolveConfigOptions
from b2c_tooling_sdk.logging import get_logger


class MobifySource:
    """Configuration source that loads the MRT API key from ``~/.mobify``."""

    name = "MobifySource"
    priority = 0

    async def load(self, options: ResolveConfigOptions) -> ConfigLoadResult | None:
        """Load the MRT API key, or ``None`` if the file/key is absent."""
        logger = get_logger("config.mobify_source")
        mobify_path = options.credentials_file or self._get_mobify_path(options.cloud_origin)

        logger.debug("[MobifySource] Checking for credentials file: %s", mobify_path)
        if not os.path.exists(mobify_path):
            logger.debug("[MobifySource] No credentials file found: %s", mobify_path)
            return None

        try:
            with open(mobify_path, encoding="utf-8") as fh:
                config = json_module.load(fh)
        except Exception as error:
            logger.debug("[MobifySource] Failed to parse credentials file %s: %s", mobify_path, error)
            raise

        api_key = config.get("api_key")
        if not api_key:
            logger.debug("[MobifySource] Credentials file found but no api_key present: %s", mobify_path)
            return None

        logger.debug("[MobifySource] Loaded credentials from %s", mobify_path)
        return ConfigLoadResult(config=NormalizedConfig(mrt_api_key=api_key), location=mobify_path)

    def _get_mobify_path(self, cloud_origin: str | None) -> str:
        """Resolve the mobify config file path based on cloud origin."""
        home = str(Path.home())
        if cloud_origin:
            hostname = urlparse(cloud_origin).hostname
            suffix = hostname if hostname else cloud_origin
            return os.path.join(home, f".mobify--{suffix}")
        return os.path.join(home, ".mobify")


__all__ = ["MobifySource"]
